### DISCUSSION TITLE
Call for `dagger run` terminal UI feedback

### DISCUSSION THREAD

**Discussion by vito** [dagger/dagger#4942] [General]
If you want to discuss this further on our Discord channel, follow this thread: https://discord.com/channels/707636530424053791/1105631313198841956

In Dagger v0.5.0 we shipped an experimental new feature: a terminal UI for `dagger run`. (https://github.com/dagger/dagger/pull/4522)

Terminal UIs are notoriously difficult to implement, so there will be quirks, and there may be times where it gets in your way. But overall it should be much better experience than `dagger.WithLogOutput(os.Stderr)`. If you run into any quirks or bugs, or have any ideas to further improve it, please leave a comment here!

![image](https://user-images.githubusercontent.com/1880/231501457-7b1350c0-35f9-46c5-827f-c4ccc16bdc53.png)

## Installation

* Grab the latest `dagger` CLI from our [GitHub releases](https://github.com/dagger/dagger/releases/tag/v0.5.0).
* Extract the `dagger` CLI from the archive and put it somewhere in your `$PATH`.
* `export _EXPERIMENTAL_DAGGER_TUI=yep` (any nonempty value).
  * I would recommend putting this in your shell profile so you can start building `dagger run` muscle memory.

## Usage

Use `dagger run` to wrap whatever command you would normally use to run Dagger:

```sh
$ dagger run go run ./dagger
```

The command will be displayed as the root of the tree view. The bottom half of the screen is dedicated to log output. The UI starts in `FOLLOW` mode by default, meaning it will automatically switch focus to actively running steps.

The terminal UI will remain open after the command exits so you have a chance to look over the results.

There are quite a few keybindings. You can press `?` to reveal them all.

* <kbd>↑</kbd> / <kbd>↓</kbd> / <kbd>Home</kbd> / <kbd>End</kbd> / <kbd>Pg Up</kbd> / <kbd>Pg Dn</kbd>
  * with tree focused: navigate the tree, and switch from `FOLLOW` mode to `BROWSE` mode
  * with logs focused: scroll the viewport
* <kbd>Tab ↹</kbd>
  * toggle focus between tree and logs
* <kbd>←</kbd> / <kbd>→</kbd>
  * collapse / expand tree entry
* <kbd>[</kbd> / <kbd>]</kbd>
  * collapse all / expand all
* <kbd>f</kbd>
  * toggle `FOLLOW` mode
* <kbd>o</kbd>
  * open tree entry in `$EDITOR`
  * if the tree entry has children, a directory will be created for each child node
  * tip: <kbd>Home</kbd> <kbd>o</kbd> to save all logs to a temporary folder and open in `$EDITOR`. Useful if you caught "lightning in a bottle."
* <kbd>q</kbd>
  * quit
  * Sends SIGTERM to the running command, followed by SIGKILL after 10 seconds
  * Let me know if this results in it getting "stuck" for 10 seconds. I personally would prefer to just SIGKILL immediately but I know it's a bit aggressive :sweat_smile: 

## Known issues

* There can sometimes be a bit of flickering as the TUI redraws. It's pretty hard to 100% fix this owing to ancient terminal technology, and a lot of optimizations have already been implemented, but definitely leave a comment if it's unusable.

**Comment by helderco** [comment]
Didn't work on www.warp.dev, but surprised that it worked fine with a [rich](https://github.com/Textualize/rich/) [status](https://github.com/dagger/dagger/blob/2ed262c629e4e589d3a335a09dcf8d098fb4848c/sdk/python/examples/client-simple/say.py#L31) 🙂 (in iTerm).

**Reply by vito** [reply]
Interesting - how does it fail? Does Warp usually work with fancy TUIs like `htop`? I would try myself but don't have a Mac handy.

**Reply by helderco** [reply]
Yes, htop works.

**Reply by helderco** [reply]
With this TUI, it just hangs (forever) like this:

<img width="196" alt="Screenshot 2023-04-12 at 18 16 41" src="https://user-images.githubusercontent.com/174525/231548423-61774b19-0444-453f-86af-200f65a478b4.png">

**Reply by kpenfound** [reply]
Is this resolved for you now on latest @helderco ? It's working for me in warp

**Reply by helderco** [reply]
Yes, it works now 👍

**Comment by helderco** [comment]
The "open logs in $EDITOR" is aweseome 🙂 A container pull with layer download and extraction opens as an empty file though.

<img width="1642" alt="Screenshot 2023-04-12 at 16 12 02" src="https://user-images.githubusercontent.com/174525/231518475-5c49a126-02e8-46c1-ac31-bb3b93e6301b.png">

Maybe that output goes into stderr and the open logs is only for stdout output?

**Reply by vito** [reply]
Oh, yeah these are actually their own special type of logs ("vertex status" :nerd_face:) so they don't end up captured like all the rest. "Open logs" should give you the interleaved stdout/stderr logs that you see in the details pane.

We can capture these logs too, just so it's less confusing. Could maybe even capture them in a more useful format, since filled progress bars don't seem all that useful at that point.

**Reply by helderco** [reply]
I see.

Yes, would be nice to see normal TTY output in real time, but then open those logs as if having run in non-TTY:

![pull-tty](https://user-images.githubusercontent.com/174525/231534623-399057ea-88da-43fd-a465-875d741769e9.gif)

![pull-non-tty](https://user-images.githubusercontent.com/174525/231534604-73b97dbb-96cd-472f-a7c2-fb82acdf0d0d.gif)

**Reply by robertu94** [reply]
A keybind for a vim like experiment might be "enter" which is what vims file bowser netrw uses to open files. 

I can also see an argument for "i" or any other binding for insert mode.

**Comment by brk3** [comment]
So far it's working great. The only issue I've found is that PgUp / PgDown doesnt seem to be working for me, possibly because I'm in tmux?

**Reply by vito** [reply]
Is that with the tree focused, the logs focused, or both? I think it might not be implemented for the logs view. :thinking:

**Reply by brk3** [reply]
Just log view, though I don’t have enough jobs right now to try it in the tree.

**Reply by jlongtine** [reply]
@vito Yeah, `PgUp` / `PgDown` doesn't work for me in log view, either. I'd love to have that, if possible. `engine:test` gets pretty boring scrolling up and down a line at a time...

**Reply by vito** [reply]
@jlongtine PR ready: https://github.com/dagger/dagger/pull/4998

**Comment by helderco** [comment]
Would be nice to collapse/expand with h/l. Or space.

**Reply by vito** [reply]
@busla just added this :tada: https://github.com/dagger/dagger/pull/4997

**Reply by robertu94** [reply]
The vim keybinding for fold/unfold is "zc" and "zo"  respectfully. "za" toggles a fold

**Comment by helderco** [comment]
Sometimes it gets stuck in the last line of "importing cache manifest from dagger:10686922502337221602". Ticker is animated and time counts up, but it can take forever unless I kill it.

<img width="874" alt="Screenshot 2023-04-14 at 14 52 21" src="https://user-images.githubusercontent.com/174525/232078614-48177e8c-9998-4f9b-b745-dcafd6f706f9.png">

**Reply by vito** [reply]
Yeah, this is a tricky one. It seems to happen with any vertex that quickly starts and finishes. I've also seen it with `mkfile`.

I attempted to debug it a couple of times, and even captured some journal logs to try and reproduce it (with `_EXPERIMENTAL_DAGGER_JOURNAL=/tmp/foo.log`), but haven't pinned it down just yet.

**Reply by 4x0v7** [reply]
Have run into the same issue intermittently

**Reply by vito** [reply]
Finally tracked this down: https://github.com/dagger/dagger/pull/4996

**Comment by helderco** [comment]
Feature request! 
- https://github.com/dagger/dagger/issues/4463

Open a shell in a new container based off of a step in the pipeline. 🚀

This demoed at one point by @aluzzardi and was just reminded of it in Community Call.

**Reply by robertu94** [reply]
Ideally there would also be a way to open the container here "after" the command fails. This likely will require some internal stuff because normally things disappear when a build step fails.

**Comment by helderco** [comment]
Another cool feature would be to open in a graph as @aluzzardi showed in https://youtu.be/bL_9GOCZy3Y?t=564

Since the tree structure doesn't correctly map to a vertex having multiple parents (dependencies), the graph would allow much better visibility to the DAG.

**Comment by vito** [comment]
I think it would be great if `dagger run` printed a summary (or full?) logs on exit. Right now when you exit out of the TUI you have nothing in your scrollback, so you can't go back and passively see where you left off, or see output you didn't know you wanted to keep at the time, etc.

Now that I've gone back and forth a bit, I found myself more comfortable with Bass's UI: it doesn't take over the whole screen, instead you stay in the shell the whole time and just focus on the bottom of the screen where it puts the currently running items. When it's done the logs are all there in your scrollback. In the current Dagger TUI the experience feels a bit more jarring and ephemeral.

![README gif](https://raw.githubusercontent.com/vito/bass/main/demos/readme.gif)

It's still nice seeing the whole tree view and interactive UI, but I think it's useful to compare/contrast. If we had it print all the output on exit I think it'd get close enough to what I want, but I'm still interested in experimenting here.

**Reply by helderco** [reply]
Yeah I though about that something like that, more like going back into the COMPLETE interactive view. But printing the logs on exit is better than nothing.

**Reply by vito** [reply]
Follow-up: there's a new inline style TUI experiment. https://github.com/dagger/dagger/pull/5078

**Comment by brk3** [comment]
I've found an issue relating to `dagger.WithWorkdir("..")` (more context [here](https://discord.com/channels/707636530424053791/1100412439356506233)).

The use case is putting your dagger go files in their own directory, e.g. "ci". I then use `dagger.WithWorkdir("..")` to set my working directory to the parent of "ci".

Take the following:

```
package main

import (
	"context"
	"fmt"
	"os"

	"dagger.io/dagger"
)

func main() {
	ctx := context.Background()

	client, err := dagger.Connect(ctx, dagger.WithWorkdir(".."), dagger.WithLogOutput(os.Stdout))
	if err != nil {
		panic(err)
	}
	defer client.Close()

	src := client.Host().Directory(".")
	ubuntu := client.Container().From("ubuntu").
		WithMountedDirectory("/work", src).
		WithWorkdir("/work").
		WithExec([]string{"ls"})

	out, err := ubuntu.Stdout(ctx)
	if err != nil {
		panic(err)
	}

	fmt.Println(out)
}

```

If I run this without the TUI (`go run main.go`), `ls` lists the files from the parent directory of where `main.go` is running. But when run from the TUI, it lists files from in "ci".

**Reply by brk3** [reply]
Found a workaround - pass the `--workdir` to `dagger`:

```
dagger --workdir .. run go run *.go
```

**Reply by sipsma** [reply]
Just FYI we have an issue tracking this bug here: https://github.com/dagger/dagger/issues/4424

**Reply by gerhard** [reply]
This is a workaround that I was looking for. Thanks @brk3!

In my case:

```console
_EXPERIMENTAL_DAGGER_TUI=1 dagger --workdir ../ run go run main.go ci
```
<img width="2880" alt="image" src="https://user-images.githubusercontent.com/3342/235232738-79f9d1c5-43c7-4f66-ba4b-0aa37a3b97ff.png">

**Comment by cpuguy83** [comment]
The TUI is really nice.

I'd really like if there were more vim-like controls.
As an example, I re-map pgUp/pgDown to my volume control because I have no other use for it (so these don't work with the TUI).
But ctrl-f/ctrl-b would be more vim like (and `less` and other TUIs).

Searching ala vim would be phenominal as well.

**Reply by robertu94** [reply]
That would be '/' for forward search and '?' For backwards search

**Reply by cpuguy83** [reply]
TIL `?`. Yeah `/` would be forward and `?` would be backwards (though I just use `n` and `N` with `/` to go one direction or the other).

**Comment by gerhard** [comment]
I am on the recently released [`v0.5.1`](https://github.com/dagger/dagger/releases/tag/v0.5.1) and the TUI doesn't seem to follow correctly.

![image](https://user-images.githubusercontent.com/3342/235232956-0d4e6f44-7b13-43d0-b74f-98e33503b632.png)

I **think** that when two ops start at once, it doesn't know which one to go to next.

To reproduce:

```console
git clone https://github.com/thechangelog/changelog.com
cd changelog.com/magefiles
_EXPERIMENTAL_DAGGER_TUI=1 dagger --workdir ../ run go run main.go ci
```

**Reply by vito** [reply]
This is definitely confusing, but probably nothing is broken. It avoids following services because usually you want to focus on the client, not the service. In this case I'm guessing the service was the last thing running, possibly waiting to just time out.

It could maybe switch to services and then switch to something else with a higher priority if it comes up.

**Reply by gerhard** [reply]
The current behaviour makes sense. I am wondering if this would have been more intuitive if services used a different colour.

**Comment by sheldonhull** [comment]
The main reason I haven't dived into using is I don't want it to keep running and require me to exit. Is that still the case? 

I want to run a task like `mage build` and it run the output until the job completes and then exit as normal.

**Reply by vito** [reply]
It still stays open, because otherwise you wouldn't be able to inspect the results, since it exits out to an empty terminal. I suggested having it print all the logs on exit in https://github.com/dagger/dagger/discussions/4942#discussioncomment-5678329 but haven't done that yet.

You might be interested in trying the inline TUI: https://github.com/dagger/dagger/pull/5078
